<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="pandoc" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
  <title>MHM User Guide, Version 2.2.0.0</title>
  <style>
    html {
      color: #1a1a1a;
      background-color: #fdfdfd;
    }
    body {
      margin: 0 auto;
      max-width: 36em;
      padding-left: 50px;
      padding-right: 50px;
      padding-top: 50px;
      padding-bottom: 50px;
      hyphens: auto;
      overflow-wrap: break-word;
      text-rendering: optimizeLegibility;
      font-kerning: normal;
    }
    @media (max-width: 600px) {
      body {
        font-size: 0.9em;
        padding: 12px;
      }
      h1 {
        font-size: 1.8em;
      }
    }
    @media print {
      html {
        background-color: white;
      }
      body {
        background-color: transparent;
        color: black;
        font-size: 12pt;
      }
      p, h2, h3 {
        orphans: 3;
        widows: 3;
      }
      h2, h3, h4 {
        page-break-after: avoid;
      }
    }
    p {
      margin: 1em 0;
    }
    a {
      color: #1a1a1a;
    }
    a:visited {
      color: #1a1a1a;
    }
    img {
      max-width: 100%;
    }
    h1, h2, h3, h4, h5, h6 {
      margin-top: 1.4em;
    }
    h5, h6 {
      font-size: 1em;
      font-style: italic;
    }
    h6 {
      font-weight: normal;
    }
    ol, ul {
      padding-left: 1.7em;
      margin-top: 1em;
    }
    li > ol, li > ul {
      margin-top: 0;
    }
    blockquote {
      margin: 1em 0 1em 1.7em;
      padding-left: 1em;
      border-left: 2px solid #e6e6e6;
      color: #606060;
    }
    code {
      font-family: Menlo, Monaco, Consolas, 'Lucida Console', monospace;
      font-size: 85%;
      margin: 0;
      hyphens: manual;
    }
    pre {
      margin: 1em 0;
      overflow: auto;
    }
    pre code {
      padding: 0;
      overflow: visible;
      overflow-wrap: normal;
    }
    .sourceCode {
     background-color: transparent;
     overflow: visible;
    }
    hr {
      background-color: #1a1a1a;
      border: none;
      height: 1px;
      margin: 1em 0;
    }
    table {
      margin: 1em 0;
      border-collapse: collapse;
      width: 100%;
      overflow-x: auto;
      display: block;
      font-variant-numeric: lining-nums tabular-nums;
    }
    table caption {
      margin-bottom: 0.75em;
    }
    tbody {
      margin-top: 0.5em;
      border-top: 1px solid #1a1a1a;
      border-bottom: 1px solid #1a1a1a;
    }
    th {
      border-top: 1px solid #1a1a1a;
      padding: 0.25em 0.5em 0.25em 0.5em;
    }
    td {
      padding: 0.125em 0.5em 0.25em 0.5em;
    }
    header {
      margin-bottom: 4em;
      text-align: center;
    }
    #TOC li {
      list-style: none;
    }
    #TOC ul {
      padding-left: 1.3em;
    }
    #TOC > ul {
      padding-left: 0;
    }
    #TOC a:not(:hover) {
      text-decoration: none;
    }
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    /* The extra [class] is a hack that increases specificity enough to
       override a similar rule in reveal.js */
    ul.task-list[class]{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      font-size: inherit;
      width: 0.8em;
      margin: 0 0.8em 0.2em -1.6em;
      vertical-align: middle;
    }
    .display.math{display: block; text-align: center; margin: 0.5rem auto;}
  </style>
  <center><h1>MHM User Guide</h1><h3>Version 2.2.0.0</h3></center>
</head>
<body>
<style type="text/css">

html {
  font-size: 100%;
  overflow-y: scroll;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
}

body {
  color: #444;
  font-family: Georgia, Palatino, 'Palatino Linotype', Times, 'Times New Roman', serif;
  font-size: 12px;
  line-height: 1.7;
  padding: 1em;
  margin: auto;
  max-width: 60em;
  background: #fefefe;
}

a {
  color: #0645ad;
  text-decoration: none;
}

a:visited {
  color: #0b0080;
}

a:hover {
  color: #06e;
}

a:active {
  color: #faa700;
}

a:focus {
  outline: thin dotted;
}

*::-moz-selection {
  background: rgba(255, 255, 0, 0.3);
  color: #000;
}

*::selection {
  background: rgba(255, 255, 0, 0.3);
  color: #000;
}

a::-moz-selection {
  background: rgba(255, 255, 0, 0.3);
  color: #0645ad;
}

a::selection {
  background: rgba(255, 255, 0, 0.3);
  color: #0645ad;
}

p {
  margin: 1em 0;
}

img {
  max-width: 100%;
}

h1, h2, h3, h4, h5, h6 {
  color: #111;
  line-height: 125%;
  margin-top: 2em;
  font-weight: normal;
}

h4, h5, h6 {
  font-weight: bold;
}

h1 {
  font-size: 2em;
}

h2 {
  font-size: 1.8em;
}

h3 {
  font-size: 1.5em;
}

h4 {
  font-size: 1.2em;
}

h5 {
  font-size: 1em;
}

h6 {
  font-size: 0.9em;
}

blockquote {
  color: #666666;
  margin: 0;
  padding-left: 3em;
  border-left: 0.5em #EEE solid;
}

hr {
  display: block;
  height: 2px;
  border: 0;
  border-top: 1px solid #aaa;
  border-bottom: 1px solid #eee;
  margin: 1em 0;
  padding: 0;
}

pre, code, kbd, samp {
  color: #000;
  font-family: monospace, monospace;
 _font-family: 'courier new', monospace;
  font-size: 0.9em;
  background-color: #EEEEEE;
}

pre {
  white-space: pre;
  white-space: pre-wrap;
  word-wrap: break-word;
}

b, strong {
  font-weight: bold;
}

dfn {
  font-style: italic;
}

ins {
  background: #ff9;
  color: #000;
  text-decoration: none;
}

mark {
  background: #ff0;
  color: #000;
  font-style: italic;
  font-weight: bold;
}

sub, sup {
  font-size: 75%;
  line-height: 0;
  position: relative;
  vertical-align: baseline;
}

sup {
  top: -0.5em;
}

sub {
  bottom: -0.25em;
}

ul, ol {
  margin: 1em 0;
  padding: 0 0 0 2em;
}

li p:last-child {
  margin-bottom: 0;
}

ul ul, ol ol {
  margin: .3em 0;
}

dl {
  margin-bottom: 1em;
}

dt {
  font-weight: bold;
  margin-bottom: .8em;
}

dd {
  margin: 0 0 .8em 2em;
}

dd:last-child {
  margin-bottom: 0;
}

#TOC ul {
    list-style-type: none;
}

img {
  border: 0;
  -ms-interpolation-mode: bicubic;
  vertical-align: middle;
}

figure {
  display: block;
  text-align: center;
  margin: 1em 0;
}

figure img {
  border: none;
  margin: 0 auto;
}

figcaption {
  font-size: 1.0em;
  font-style: bold;
  margin: 0 0 .8em;
}

table {
  margin-bottom: 2em;
  border-bottom: 1px solid #ddd;
  border-right: 1px solid #ddd;
  border-spacing: 0;
  border-collapse: collapse;
}

table th {
  padding: .2em 1em;
  background-color: #eee;
  border-top: 1px solid #ddd;
  border-left: 1px solid #ddd;
}

table td {
  padding: .2em 1em;
  border-top: 1px solid #ddd;
  border-left: 1px solid #ddd;
  vertical-align: top;
}

.author {
  font-size: 1.2em;
  text-align: center;
}

@media only screen and (min-width: 480px) {
  body {
    font-size: 14px;
  }
}
@media only screen and (min-width: 768px) {
  body {
    font-size: 16px;
  }
}


@media print {
  {
    background: transparent !important;
    color: black !important;
    filter: none !important;
    -ms-filter: none !important;
  }

  body {
    font-size: 10pt;
    max-width: 100%;
  }

  title {
      font-size: 2em;
  }
  
  a, a:visited {
      text-decoration: none;
      color: #520;
  }

  #TOC {
      page-break-after: always;
      font-size: 12pt;
  }

  #introduction {
      page-break-before: always;
  }
  
  hr {
    height: 1px;
    border: 0;
    border-bottom: 1px solid black;
  }

  a[href]:after {
    content: " (" attr(href) ")";
  }

  abbr[title]:after {
    content: " (" attr(title) ")";
  }

  .ir a:after, a[href^="javascript:"]:after, a[href^="#"]:after {
      content: "";
  }

  pre, blockquote {
      padding-right: 1em;
      padding-left: 2em;
      page-break-inside: avoid;
  }

  tr, img {
    page-break-inside: avoid;
  }

  img {
    max-width: 100% !important;
  }

  @page {
    margin: 15mm 15mm 15mm 15mm;
  }

  p, h2, h3 {
    orphans: 3;
    widows: 3;
  }

  h1, h2, h3 {
      page-break-after: avoid;
  }

  @page :footer {
      content: "Page " decimal(pageno);
      font-variant: small-caps;
  }
}


</style>
<hr />
<p><strong><em>Copyright</em></strong></p>
<p><em>This manuscript has been authored by an author at Lawrence
Berkeley National Laboratory under Contract No. DE-AC02-05CH11231 with
the U.S. Department of Energy. The U.S. Government retains, and the
publisher, by accepting the article for publication, acknowledges, that
the U.S. Government retains a non-exclusive, paid-up, irrevocable,
world-wide license to publish or reproduce the published form of this
manuscript, or allow others to do so, for U.S. Government
purposes.</em></p>
<p><strong><em>Legal Disclaimer</em></strong></p>
<p><em>This document was prepared as an account of work sponsored by the
United States Government. While this document is believed to contain
correct information, neither the United States Government nor any agency
thereof, nor the Regents of the University of California, nor any of
their employees, makes any warranty, express or implied, or assumes any
legal responsibility for the accuracy, completeness, or usefulness of
any information, apparatus, product, or process disclosed, or represents
that its use would not infringe privately owned rights. Reference herein
to any specific commercial product, process, or service by its trade
name, trademark, manufacturer, or otherwise, does not necessarily
constitute or imply its endorsement, recommendation, or favoring by the
United States Government or any agency thereof, or the Regents of the
University of California. The views and opinions of authors expressed
herein do not necessarily state or reflect those of the United States
Government or any agency thereof or the Regents of the University of
California.</em></p>
<p><strong><em>Acknowledgments</em></strong></p>
<p><em>This work was supported by the Exascale Computing Project
(17-SC-20-SC), a collaborative effort of the U.S. Department of Energy
Office of Science and the National Nuclear Security
Administration.</em></p>
<hr />
<p><a
href="https://sites.google.com/lbl.gov/exabiome/downloads?authuser=0">MetaHipMer
(<em>MHM</em>)</a> is a <em>de novo</em> metagenome short-read
assembler. This is version 2 (MHM2), which is written in <a
href="https://upcxx.lbl.gov">UPC++</a>, CUDA and HIP, and runs
efficiently on both single servers and on multinode supercomputers,
where it can scale up to coassemble terabase-sized metagenomes. More
information about MetaHipMer can be found under the <a
href="https://sites.google.com/lbl.gov/exabiome">ExaBiome Project</a> of
the <a href="https://www.exascaleproject.org">Exascale Computing
Project</a> and in several publications:</p>
<ul>
<li><a href="https://ieeexplore.ieee.org/document/8665813">E. Georganas
et al., “Extreme Scale De Novo Metagenome Assembly,” SC18: International
Conference for High Performance Computing, Networking, Storage and
Analysis, Dallas, TX, USA, 2018, pp. 122-13.</a></li>
<li><a
href="https://www.nature.com/articles/s41598-020-67416-5#citeas">Hofmeyr,
S., Egan, R., Georganas, E. et al. Terabase-scale metagenome coassembly
with MetaHipMer. Sci Rep 10, 10689 (2020).</a></li>
<li><a
href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-020-03720-1#citeas">Awan,
M.G., Deslippe, J., Buluc, A. et al. ADEPT: a domain independent
sequence alignment strategy for gpu architectures. BMC Bioinformatics
21, 406 (2020).</a></li>
<li><a href="https://dl.acm.org/doi/pdf/10.1145/3458817.3476212">Muaaz
Awan, Steven Hofmeyr, Rob Egan et al. “Accelerating large scale de novo
metagenome assembly using GPUs.”, SC 2021</a></li>
</ul>
<p>The quality of MetaHipMer assemblies is comparable with other leading
metagenome assemblers, as documented in the results of the CAMI2
competition, where MetaHipMer was rated first for quality in two out of
three datasets, and second in the third dataset:</p>
<ul>
<li><a href="https://www.nature.com/articles/s41592-022-01431-4">F.
Meyer et al., “Critical Assessment of Metagenome Interpretation: the
second round of challenges”, Nature Methods volume 19, pages429–440
(2022)</a></li>
</ul>
<p>MHM2 is developed and maintained by the <a
href="https://sites.google.com/lbl.gov/exabiome/">Exabiome Project</a>
at Lawrence Berkeley National Laboratory, and is supported by the <a
href="https://www.exascaleproject.org">Exascale Computing Project</a>
(17-SC-20-SC), a collaborative effort of the U.S. Department of Energy
Office of Science and the National Nuclear Security Administration.</p>
<h1 id="building-and-installing">Building and Installing</h1>
<p>MHM2 depends on UPC++, with the C++17 standard, and CMake. GPU builds
require CUDA and/or HIP.</p>
<p>A script, <a
href="https://bitbucket.org/berkeleylab/mhm2/src/master/build.sh"><code>build.sh</code></a>,
is provided for building and installing MHM2.</p>
<p>Before building MHM2, ensure that either the UPC++ compiler wrapper,
<code>upcxx</code> is in your <code>PATH</code>, or set the
<code>MHM2_BUILD_ENV</code> environment variable to point to a script
that loads the appropriate environment, for example, on <a
href="https://docs.nersc.gov/systems/perlmutter">NERSC’s Perlmutter
supercomputer</a>, you would set the following for the gnu compiler on
the KNL partition:</p>
<p><code>export MHM2_BUILD_ENV=contrib/environments/perlmutter/gnu.sh</code></p>
<p>There are several scripts provided for different build choices on
NERC’s and OLCF’s systems, in directories that start with
<code>contrib/environments</code>. You do not need to use any scripts
such as these when building on a Linux server, although you may want to
create your own when setting up the build. On NERSC and OLCF we
recommend using the gnu (<code>contrib/environments/*/gnu.sh</code>)
environments. Building with Intel is very slow.</p>
<p>To build a release version (optimized for performance), execute:</p>
<p><code>./build.sh Release</code></p>
<p>Alternatively, you can build a debug version with:</p>
<p><code>./build.sh Debug</code></p>
<p>This will capture a great deal of information useful for debugging
but will run a lot slower (up to 5x slower at scale on multiple
nodes).</p>
<p>An alternative to the pure debug version is the “release” debug
version, which still captures a reasonable amount of debugging
information, but is a lot faster (although still up to 2x slower than
the release version):</p>
<p><code>./build.sh RelWithDebInfo</code></p>
<p>The <code>./build.sh</code> script will install the binaries by
default into the <code>install/bin</code> subdirectory in the repository
root directory. To set a different install directory, set the
environment variable <code>MHM2_INSTALL_PATH</code>, e.g.:</p>
<p><code>MHM2_INSTALL_PATH=/usr/local/share/mhm2 ./build.sh Release</code></p>
<p>Once MHM2 has been built once, you can rebuild with</p>
<p><code>./build.sh</code></p>
<p>and it will build using the previously chosen setting
(<code>Release</code>, <code>Debug</code>, or
<code>RelWithDebInfo</code>).</p>
<p>You can also run</p>
<p><code>./build.sh clean</code></p>
<p>to start from scratch. If you run this, then the next call to
<code>build.sh</code> should be with one of the three configuration
settings.</p>
<p>By default, the build occurs within the root of the repository, in a
subdirectory called <code>.build</code>. This is created automatically
by the <code>build.sh</code> script.</p>
<p>The MHM2 build uses <code>cmake</code>, which you can call directly,
instead of through the <code>build.sh</code> script, e.g.:</p>
<pre><code>mkdir -p .build
cd .build
cmake -DCMAKE_INSTALL_PREFIX=path-to-install ..
make -j all install</code></pre>
<p>Consult the <code>build.sh</code> script to see how it executes these
commands.</p>
<p>You’ll need to first set the environment, e.g.:</p>
<p><code>source contrib/environments/perlmutter/gnu.sh</code></p>
<p>If you see an error message when building like the following:</p>
<p><code>include could not find load file: GetGitVersion</code></p>
<p>Then you have probably not cloned the git submodules. You need to
execute the following from the root directory:</p>
<pre><code>git submodule init
git submodule update</code></pre>
<h1 id="running">Running</h1>
<p>To execute MHM2, run the <code>mhm2.py</code> script located at
<code>install/bin</code>. Most parameters have sensible defaults, so it
is possible to run with only the read FASTQ files specified, e.g. to run
with two interleaved reads files, <code>lib1.fastq</code> and
<code>lib2.fastq</code>, you could execute:</p>
<p><code>mhm2.py -r lib1.fastq,lib2.fastq</code></p>
<p>A list of all the command line options can be found by running with
<code>-h</code>. Because <code>mhm2.py</code> is a python script that
wraps the UPC++ binary, <code>mhm2</code>, there will be two levels of
options, one from the python script, and one from the binary. Some of
the options have a short form (a single dash with a single character)
and a long form (starting with a double-dash). In the options described
below, where both a short form and a long form exist, they are separated
by a comma. The type of the option is indicated as one of
<code>STRING</code> (a string of characters), <code>INT</code> (an
integer), <code>FLOAT</code> (a floating point value) or
<code>BOOL</code> (a boolean flag). For <code>BOOL</code>, the option
can be given as <code>true</code>, <code>false</code>, <code>yes</code>,
<code>no</code>, <code>0</code>, <code>1</code>, or omitted altogether,
in which case the option will be <code>true</code>, and if an option is
specfied, the <code>=</code> must be used, e.g.</p>
<p><code>mhm2.py --checkpoint=false</code></p>
<p>By default, the run will generate files in a specific output
directory (see the <code>--output</code> option below). At a minimum,
this will include the following files:</p>
<ul>
<li><code>final_assembly.fasta</code>: the contigs for the assembly, in
FASTA format.</li>
<li><code>mhm2.log</code>: a log file containing details about the run,
including various quality statistics, details about the assembly process
and timing information.</li>
<li><code>mhm2.config</code>: a configuration file containing all the
non-default options used for the run.</li>
<li><code>per_thread</code>: a subdirectory containing per-process files
that record memory usage and debugging information in Debug mode.</li>
</ul>
<p>In addition, many more files may be generated according to which
command-line options are specified. These are described in detail below
where relevant.</p>
<p>The <code>mhm2</code> binary can be executed directly using
<code>upcxx-run</code>, <code>srun</code> or another suitable launcher.
Generally we recommend using <code>mhm2.py</code>, because it takes care
of many facets of starting the executable in a given environment and
provides additional functionality, e.g. automatically restarting on
errors, easily enabling communication tracing, etc.</p>
<p>A description of all the available options is provided below. This
list can be obtained in an abbreviated form by running
<code>mhm2.py -h</code>.</p>
<h2 id="options-provided-by-mhm2.py">Options provided by mhm2.py</h2>
<p>The following options are provided by the <code>mhm2.py</code>
wrapper script:</p>
<p><strong><code>--auto-resume BOOL</code></strong> Automatically resume
a run if it fails. This wil restart the run at the last saved
checkpoint. This defaults to false.</p>
<p><strong><code>--shared-heap INT</code></strong></p>
<p>Set the shared heap size used by the UPC++ runtime, as a percentage
of the available memory. This defaults to 10% and should not need to be
adjusted. If MHM2 fails with <code>upcxx::bad_shared_alloc</code>
messages, then this value should be increased.</p>
<p><strong><code>--procs INT</code></strong></p>
<p>Set the number of processes used when running MHM2. By default, MHM2
automatically detects the number of available cores and runs with one
process per core. This setting allows a user to run MHM2 on a subset of
available processors, which may be desirable if running on a server that
is running other applications.</p>
<p><strong><code>--nodes INT</code></strong></p>
<p>Set the number of nodes on which to execute MHM2. By default, MHM2
will use the total available number of nodes from the job launch. On
rare occasions it may be desirable to explicitly set this through the
MHM2 launch, and not during the external job launch.</p>
<p><strong><code>--gasnet-stats</code></strong></p>
<p>Collect GASNet communication statistics. This option should be run
with either a <code>Debug</code> or a <code>RelWithDebInfo</code> build.
It will produce a summary of communication statistics for each stage,
and write the summary to the log file and to stdout. Enabling this will
further increase the runtime.</p>
<p><strong><code>--gasnet-trace</code></strong></p>
<p>Enable GASNet tracing. This must be run with a <code>Debug</code> or
<code>RelWithDebInfo</code> build. This will produce one file per
process <code>P</code>, called <code>trace_&lt;P&gt;.txt</code>. For
more about what data are collected, consult the <a
href="https://gasnet.lbl.gov/dist-ex/README">GASNet documentation</a>,
in the section about GASNet tracing and statistical collection. The
trace mask is set to a comprensive set of default, i.e. setting the
GASNet environment variable: <code>GASNET_TRACEMASK="GPWBNIH"</code>.
This can be overriden by explicitly setting that environment
variable.</p>
<p><strong><code>--preproc STRING</code></strong></p>
<p>Preprocessing commands. This is a comma separated list containing
preprocesses and/or options
(e.g. <code>valgrind --leak-check=full</code>), or additional options to
be passed to <code>upcxx-run</code>.</p>
<p><strong><code>--binary STRING</code></strong></p>
<p>The name of the binary file for executing MHM2. This defaults to
<code>mhm2</code>. This option can be used to run different binaries
from the same script, such as builds with and without GPU support.</p>
<h2 id="basic-options">Basic options</h2>
<p>These are the most commonly used options.</p>
<p>The input files of reads are specified with either <code>-r</code>,
<code>-p</code>, or <code>-u</code>. At least one of these options must
be specified. When running on a <a href="https://lustre.org/">Lustre</a>
file system (such as on OLCF’s Frontier), it is recommended that all
input files be <a
href="https://wiki.lustre.org/Configuring_Lustre_File_Striping">striped</a>
to ensure adequate I/O performance. Usually this means first striping a
directory and then moving files into it, e.g. for a file
<code>reads.fastq</code>:</p>
<pre><code>mkdir data
lfs setstripe -c 72 data
mv reads.fastq data</code></pre>
<p><strong><code>-r, --reads STRING,STRING,...</code></strong></p>
<p>A collection of names of files containing interleaved paired reads in
FASTQ format. Multiple files must be comma-separated, or can be
separated by spaces. For paired reads in separate files, use the
<code>-p</code> option. For unpaired reads, use the <code>-u</code>
option. Long lists of read files can be set in a configuration file and
loaded with the <code>--config</code> option, to avoid having to type
them in on the command line.</p>
<p><strong><code>-p, --paired-reads STRING,STRING,...</code></strong></p>
<p>A collection of names of files containing separate paired reads in
FASTQ format. Multiple files must be comma-separated, or can be
separated by spaces. For each library, the file containing the reads for
the first pairs must be followed by the file containing the reads for
the second pairs, e.g. for two libraries with separate paired reads
files <code>lib1_1.fastq</code>, <code>lib1_2.fastq</code> and
<code>lib2_1.fastq</code>, <code>lib2_2.fastq</code>, the option should
be specified as:</p>
<p><code>-p lib1_1.fastq,lib1_2.fastq,lib2_1.fastq,lib2_2.fastq</code></p>
<p>This option only supports reads where each pair of reads has the same
sequence length, usually only seen in raw reads. For support of trimmed
reads of possibly different lengths, first interleave the files and then
call with the <code>-r</code> option. The separate files can be
interleaved with <code>reformat.sh</code> from <a
href="https://jgi.doe.gov/data-and-tools/bbtools/">bbtools</a>.</p>
<p><strong><code>-u, --unpaired-reads STRING,STRING,...</code></strong></p>
<p>A collection of names of files containing unpaired reads in FASTQ
format. Multiple files must be comma-separated, or can be separated by
spaces.</p>
<p><strong><code>-k, --kmer-lens INT,INT,...</code></strong></p>
<p>The <em>k</em>-mer lengths used for the contigging rounds. MHM2
performs one or more contigging rounds, each of which performs
<em>k</em>-mer counting, followed by a deBruijn graph traversal, then
alignment and local assembly to extend the contigs. Typically, multiple
rounds are used with increasing values of <em>k</em>; the shorter values
are useful for low abundance genomes, whereas the longer <em>k</em>
values are useful for resolving repeats. This option defaults to
<code>-k 21,31,47,63,95</code> for reads of average length at least 130,
and to <code>-k 21,31,47,63</code> for shorter reads. For reads longer
than 150, it may be a good idea to adjust these values, for example,
adding 121. Also, each round of contigging takes time, so the overall
assembly time an be reduced by reducing the number of rounds, although
this will likely reduce the quality of the final assembly.</p>
<p><strong><code>-s, --scaff-kmer-lens INT,INT,...</code></strong></p>
<p>The <em>k</em>-mer lengths used for the scaffolding rounds. In MHM2,
the contigging rounds are followed by one or more scaffolding rounds.
These rounds usually proceed from a high <em>k</em> to a low one,
i.e. the reverse ordering of contigging. This option defaults to
<code>-s 95,47,31</code> for reads of length greater than 130, and to
<code>-s 47,31</code> for shorter reads. The first value should always
be set to the final <code>k</code> used in contigging, e.g. for reads of
length 250 with parameter <code>-k 21,33,47,63,95,121</code>, the
scaffolding values could be <code>-s 121,63,31</code>. More rounds may
improve contiguity but will likely increase misassemblies. To disable
scaffolding altogether, set this value to 0, i.e. <code>-s 0</code>.</p>
<p><strong><code>--min-ctg-print-len INT</code></strong></p>
<p>The minimum length for contigs to be included in the final assembly,
<code>final_assembly.fasta</code>. This defaults to 500.</p>
<p><strong><code>-o, --output STRING</code></strong></p>
<p>The name for the output directory. If not specified, it will be set
to a default value of the following form:</p>
<p><code>mhm2-run-&lt;READS_FNAME1&gt;-n&lt;PROCS&gt;-N&lt;NODES&gt;-YYMMDDhhmmss-&lt;JOBID&gt;</code></p>
<p>where <code>&lt;READS_FNAME1&gt;</code> is the name of the first
reads file, <code>PROCS</code> is the number of processes and
<code>NODES</code> is the number of nodes. Following this is the date
and time when the run was started: <code>YY</code> is the last two
digits of the year, <code>MM</code> is the number of the month,
<code>DD</code> is the day of the month, <code>hh</code> is the hour of
day, <code>mm</code> is the minute and <code>ss</code> is the second. Be
warned that if two runs are started at exactly the same time, with the
same parameters, then with the default values, they could both end up
running in the same output directory, which will lead to corrupted
results.</p>
<p>If the output directory is created by MHM2 (either as the default or
when passed as a parameter), it will automatically be striped in the
most effective way on a Lustre filesystem. If using a pre-existing
directory that was not created by MHM2, the user should ensure that on
Lustre filesystems it is adequately striped.</p>
<p>If the output directory already exists, files produced by a previous
run of MHM2 may be overwritten, depending on whether or not this is a
restart of a previous run. If there is an existing log file
(<code>mhm2.log</code>), it will be renamed with the date appended
before the new one is written as <code>mhm2.log</code>, so log
information about previous runs will always be retained.</p>
<p><strong><code>--checkpoint BOOL</code></strong></p>
<p>Checkpoint runs. If set to true, this will checkpoint the run by
saving intermediate files that can later be used to restart the run (see
the <code>--restart</code> option below). The intermediate files are
FASTA files of contigs, and they are saved at the end of each contigging
round (<code>contigs-&lt;k&gt;.fasta</code>) and at the end of each
scaffolding round (<code>scaff-contigs-&lt;k&gt;.fasta</code>), where
the <code>&lt;k&gt;</code> value is the <em>k</em>-mer size for that
round. Checkpointing is on by default and can be disabled by passing
<code>--checkpoint=false</code>.</p>
<p><strong><code>--restart BOOL</code></strong></p>
<p>Restart a previous incomplete run. If set to true, MHM2 will attempt
to restart a run from an existing directory. The output directory option
must be specified and must contain a previous checkpointed run. The
restart will use the same options as the previous run, and will load the
most recent checkpointed contigs file in order to resume. This defaults
to false.</p>
<p><strong><code>--post-asm BOOL</code></strong></p>
<p>Perform alignment of reads to final assembly after assembly has
completed, and compute abundances. If set to true. MHM2 will align the
original reads to the final assembly and write one output SAM file (with
extension <code>.sam</code>) per input file. The headers for all the SAM
files will be all written to a single file,
<code>final_assembly.header.sam</code>. The output will be in <a
href="https://samtools.github.io/hts-specs/SAMv1.pdf">SAM format</a>.
MHM2 will also compute the abundances (depths) for the scaffolds in the
final assembly and write the results to the file,
<code>final_assembly_depths.txt</code>. The format of this file is the
same as that used by <a
href="https://bitbucket.org/berkeleylab/metabat/">MetaBAT</a>, and so
can be used together with the <code>final_assembly.fasta</code> for
post-assembly binning, e.g.:</p>
<p><code>metabat2 -i final_assembly.fasta -a final_assembly_depths.txt -o bins_dir/bin</code></p>
<p>This defaults to false.</p>
<p><strong><code>--post-asm-only BOOL</code></strong></p>
<p>Perform only post-assembly operations. If set to true, this requires
an existing directory containing a full run (i.e. with a
<code>final_assembly.fasta</code> file), and it will execute
post-assembly operations on that assembly without any other steps. This
provides a convenient means to run alignment and/or abundance
calculations on an already completed assembly. By default this
post-assembly analysis will use the <code>final_assembly.fasta</code>
file in the output directory, but any FASTA file could be used,
including those not generated by MHM2 (see the <code>--contigs</code> in
the advanced options section below). This defaults to false.</p>
<p><strong><code>--post-asm-write-sam BOOL</code></strong></p>
<p>Turn writing of the SAM files for post-assembly on or off. This
defaults to true. The SAM files can be very large and do not need to be
saved to the filesystem in order for the abundances to be calculated,
and only the abundances are needed for running MetaBAT. Specify false
here to save space.</p>
<p><strong><code>--progress BOOL</code></strong></p>
<p>Display progress indicators during a run. If true, many
time-consuming stages will be shown updating with a simple progress bar.
The progress bar output will not be written into the log file,
<code>mhm2.log</code>. This defaults to false.</p>
<p><strong><code>-v, --verbose BOOL</code></strong></p>
<p>Verbose output. If true, MMHM2 will produce verbose output, which
prints out a lot of additional information about the timing of the run
and the various computations that are being performed. This defaults to
false. All of the information seen in verbose mode will always be
written to the log file, <code>mhm2.log</code>. This defaults to
false.</p>
<p><strong><code>--config STRING</code></strong></p>
<p>Use a config file for the parameters. If this is specified, the
options will be loaded from the named config file. The file is a plain
text file of the format:</p>
<p><code>key = value</code></p>
<p>where <code>key</code> is the name of an option and
<code>value</code> is the value of the option. All blank lines and lines
beginning with a semi-colon will be ignored. When the config file is not
specified as an option, MHM2 always writes out all of the non-default
options to the file <code>mhm2.config</code> in the output directory.
Even when options are loaded from a config file, they can still be
overridden by options on the command line. For example, if the config
file, <code>test.config</code>, contains the line:</p>
<p><code>k = 21,33,55,77,99</code></p>
<p>but the command line is:</p>
<p><code>mhm2.py --config test.config -k 45,63</code></p>
<p>then MHM2 will run with <em>k</em>-mer lengths of 45, 63.</p>
<h2 id="advanced-options">Advanced options</h2>
<p>These are additional options for tuning performance or the quality of
the output, selecting precisely how to restart a run, or for additonal
debugging information. Most users will not need any of these
options.</p>
<h3 id="restarting-runs">Restarting runs</h3>
<p>Although the <code>--restart</code> option provides for simple
restarts of previous runs, it is possible to restart at very specific
points, with different options from those of the original run,
e.g. restarting scaffolding with different <em>k</em>-mer values, set
using the <code>-s</code> option.</p>
<p>The relevant options are listed below.</p>
<p><strong><code>-c, --contigs STRING</code></strong></p>
<p>The file name containing contigs in FASTA format that are to be used
as the most recent checkpoint for a restart. Any contigs file generated
during a checkpointed run can be used, so it is possible to restart at
any stage. It is also possible to specify <em>any</em> FASTA file if
running only post-assembly analysis (<code>--post-asm-only</code>).</p>
<p><strong><code>--max-kmer-len INT</code></strong></p>
<p>The maximum <em>k</em>-mer length that was previously used in
contigging. This is usually derived from the <code>-k</code> parameter,
and so only needs to be specified if the restart will run scaffolding
rounds only. For example, the following command will restart after the
scaffolding round with <code>k=99</code> and will run run two more
scaffolding rounds with <code>k=55</code> and <code>k=21</code>:</p>
<p><code>mhm2.py -o outdir -r reads.fq -c scaff-contigs-99.fasta --max-kmer-len 99 -s 55,21</code></p>
<p><strong><code>--prev-kmer-len INT</code></strong></p>
<p>The <em>k</em>-mer length in the previous contigging round. Only
needed if restarting in contigging, e.g.</p>
<p><code>mhm2.py -o outdir -r reads.fq -c contigs-77.fasta --max-prev-kmer-len 55</code></p>
<h3 id="tuning-assembly-quality">Tuning assembly quality</h3>
<p>There are several additonal options for adjusting the quality of the
final assembly, apart from the <em>k</em>-mer values specified for the
contigging and scaffolding rounds, as described earlier.</p>
<p><strong><code>--adapter-trim BOOL</code></strong></p>
<p>If set to true, will trim adapters using the
<code>--adapter-refs</code> file. This can be turned off if the input
reads are already trimmed and preprocessed to removed adapters. The
default is true.</p>
<p><strong><code>--adapter-refs STRING</code></strong></p>
<p>A file containing adapter sequences in the FASTA format. If
specified, it will be used to trim out all adapters when the input reads
are first loaded. Two files containing adapter sequences are provided in
the <code>contrib</code> directory:
<code>adapters_no_transposase.fa</code> and
<code>all_adapters.fa.gz</code>. When installed the latter is unzipped
and copied to the install directory under <code>share</code>. The
default is to trim adapters using this file, so this option usually does
not need to be set.</p>
<p><strong><code>-i, --insert INT:INT</code></strong></p>
<p>The insert size for paired reads. The first integer is the average
insert size for the paired reads and the second integer is the standard
deviation of the insert sizes. MHM2 will automatically attempt to
compute these values so this parameter is usually not necessary.
However, there are certain cases where it may be useful, for example, if
MHM2 prints a warning about being unable to compute the insert size
because of the nature of the reads, or if only doing scaffolding. MHM2
will also compare its computed value to any option set on the command
line and print a warning if the two differ significantly; this is useful
for confirming assumptions about the insert sise distribution.</p>
<p><strong><code>--break-scaff-Ns INT</code></strong></p>
<p>The number of Ns allowed in a gap before the scaffold is broken into
two. The default is 10.</p>
<p><strong><code>--min-depth-thres INT</code></strong></p>
<p>The minimum depth (abundance) for a <em>k</em>-mer to be considered
for the deBruijn graph traversal. This defaults to 2. Increasing it can
reduce errors at the cost of reduced contiguity and genome
discovery.</p>
<p><strong><code>--optimize STRING</code></strong></p>
<p>Adjust the trade-off in assembly quality between errors and
contiguity. There are three settings that can be used:
<code>contiguity</code> (improve contiguity at the cost of increased
errors), <code>correctness</code> (reduce errors at the cost of
contiguity), and <code>default</code> (the default setting which tries
to balance the two).</p>
<h3 id="adjusting-performance-and-memory-usage">Adjusting performance
and memory usage</h3>
<p>There are several options that adjust the trade-off between the
memory used and the time taken, or that influence the performance in
different ways on different platforms. These usually do not have to be
adjusted.</p>
<p><strong><code>--max-kmer-store INT</code></strong></p>
<p>The maximum size per process in MB for the aggregation of
<em>k</em>-mers during <em>k</em>-mer analysis. This defaults to 1% of
available memory. Higher values use more memory, but can potentially
result in faster computation.</p>
<p><strong><code>--max-rpcs-in-flight INT</code></strong></p>
<p>The maximum number of remote procedure calls (RPCs) outstanding at
any given time. The default is 100. Reducing this will reduce memory
usage but could increase running time. If set to 0, there are no limits
on the outstanding RPCs.</p>
<p><strong><code>--aln-ctg-seq-buf-size INT</code></strong></p>
<p>Size of buffer for fetching contig sequences during alignment.
Reducing this will reduce memory usage but slow down communication.
Defaults to 1000.</p>
<p><strong><code>--max-worker-threads INT</code></strong></p>
<p>The maximum number of background worker threads. These threads are
used for a limited number of background computation tasks. The default
value is three and should not need to be adjusted.</p>
<p><strong><code>--pin STRING</code></strong></p>
<p>Restrict the hardware contexts that processes can run on. There are
five options: <code>cpu</code> (restrict each process to a single
logical CPU); <code>core</code> (restrict each process to a core);
<code>numa</code> (restrict each process to a NUMA domain);
<code>rr_numa</code> (restrict each process to a NUMA domain in a round
robin manner; and <code>none</code> (don’t restrict the processes). The
default is <code>cpu</code>.</p>
<p><strong><code>--sequencing-depth INT</code></strong></p>
<p>The expected average sequencing depth. This value is used to estimate
the memory requirements for unique <em>k</em>-mers in the first round of
contigging. It may only need to be adjusted when memory is scarce and
there is a need to better balance initial memory allocations.</p>
<p><strong><code>--post-asm-subsets INT</code></strong></p>
<p>Number of subsets to split contigs into for post-assembly. Each
subset is processed in turn against all input files and the results are
combined per input file. Increasing the number of subsets reduces memory
usage but increases time taken.</p>
<p><strong><code>--shuffle-reads BOOL</code></strong></p>
<p>Shuffle reads to improve locality. This defaults to true, and results
in greatly enhanced performance. The only reason to disable it is when
testing or evaluating the impact of read shuffling.</p>
<p><strong><code>--use-qf BOOL</code></strong></p>
<p>Use the <a
href="https://ppopp23.sigplan.org/details/PPoPP-2023-papers/13/High-Performance-Filters-for-GPUs">TCF
filter</a> to reduce memory when running on GPUs. This option will
reduce peak GPU memory requirements by about 50%, with a performance
impact of less than one percent. The only reason to disable it is for
debugging or evaluating the impact of the TCF.</p>
<p><strong><code>--subsample-pct INT</code></strong></p>
<p>Percentage of input read files to use in the assembly. The value is
from 0 to 100. This option enables to user to test a dataset with a
smaller part of it, e.g. 10%. All of the input files will be reduced to
the percentage specified.</p>
<h3 id="miscellaneous">Miscellaneous</h3>
<p><strong><code>--dump-merged BOOL</code></strong></p>
<p>Write merged FASTQ input files to the output directory. The file will
be named <code>&lt;READ_FILE_NAME&gt;-merged.fastq</code>. The default
is false.</p>
<p><strong><code>--dump-kmers BOOL</code></strong></p>
<p>Write <em>k</em>-mers to files in the output directory. The
<em>k</em>-mers are written in each contigging stage immediately after
<em>k</em>-mer counting. The files will be written on a per process
basis into the <code>per_rank</code> subdirectory, named
<code>kmers-&lt;k&gt;.txt.gz</code>, where <code>k</code> is the value
for the contigging round. The default is false.</p>
<p><strong><code>--write-gfa BOOL</code></strong></p>
<p>Produce an assembly graph in the GF2 format. If set to true, MHM2
will output an assembly graph in the <a
href="https://github.com/GFA-spec/GFA-spec/blob/master/GFA2.md">GFA2
format</a> in the file, <code>final_assembly.gfa</code>. This represents
the assembly graph formed by aligning the reads to the final contigs and
using those alignments to infer edges between the contigs. This defaults
to false.</p>
<p><strong><code>-Q, --quality-offset INT</code></strong></p>
<p>The <em>phred</em> encoding offset. In most cases, MHM2 will be able
to detect this offset from analyzing the reads file, so it usually does
not need to be explicitly set.</p>
</body>
</html>
